src/notification.cpp

```diff
@@ -1190,13 +1228,57 @@ void Notification::checkActionInvoked(uint id, QString actionKey)
 {
     Q_D(Notification);
     if (id == d->replacesId) {
+        // TODO: check reply is filled in for actions that require them
+        foreach (const QVariant &action, d->remoteActions) {
+            QVariantMap vm = action.value<QVariantMap>();
+            const QString actionName = vm["name"].value<QString>();
+            if (!actionName.isEmpty() && actionName == actionKey) {
+                if ((vm.contains("replyChoices") || vm.contains("replyEditable")) // Need reply
+                    && (d->replyText.isEmpty() // Reply is empty
+                        || (vm["replyEditable"].toBool() == false // or not editable 
+                            && !(vm["replyChoices"].value<QStringList>().contains(d->replyText))
+                            // and not a valid choice 
```

Comment:
Who's to decide if empty text is valid? Could be even an option to delegate all validness checks to the app?

Reply:
I'm not really very happy with a return with no feedback happening with invalid data. Delegating this to the app would mean it could issue another notification or otherwise alert the user when invalid data was supplied, but if it's supplied choices here so it's reasonable to expect that to be validated. Worried about the scope of this expanding too much when it's really up to the notification displayer to validate.

We possibly also need some sort of easy way for apps to check if notification action replies are possible on the version they're running on, so they can omit actions requiring them.

Reply:
Still: who's to decide what is valid or invalid data? Don't have use-case immediately, but could be that sometimes an empty string is ok for the app?

For the case of choices given but editable set false suppose that can be considered invalid.

Reply:
I think the rare occasion where a reply would be optional would be outweighed by the risk of someone accidentally sending an empty message. Maybe allowing empty strings would require app validation.